### PR TITLE
Refactor codebase

### PR DESCRIPTION
**Pull request by ManyTheFish**
# Summary
Big refactor of the codebase.

This PR contains only main structures and traits, and some implemented examples of Segmenter and Normalizer to showcase the shape of the implementation.

The other Segmenters and Normalizers will be implemented in the next PRs.
The tests and benchmarks will be added/modified in the next PRs.

# Goal

The goal of this PR is to reshape the tokenizer to ease the addition of a new `Segmenter` and to add documentation of the library to release a public crate in the future.

Because there is a lot of breaking changes, I recommend browsing directly directories instead of looking at diffs.

Please run `cargo doc --open` and give some feedback about the final rendering of the documentation and the used naming, thanks!

related to #72 

note that this PR will not be merged on the main branch but on an intermediate one.

### CODE REVIEW

File: src/segmenter/mod.rs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        })
+    }
+}
+
+/// Try to Detect Language and Script and return the corresponding segmenter,
+/// if no Language is detected or no segmenter corresponds to the Language
+/// the function try to get a segmenter corresponding to the script;
+/// if no Script is detected or no segmenter corresponds to the Script,
+/// the function try to get the default segmenter in the map;
+/// if no default segmenter exist in the map return the librairy DEFAULT_SEGMENTER.
```
Comment (irevoire) [inline src/segmenter/mod.rs:70]:
```suggestion
/// if no default segmenter exist in the map return the library DEFAULT_SEGMENTER.
```

File: src/tokenizer.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    fn tokenize(&self) -> ClassifiedTokenIter<'_, '_, Vec<u8>> {
+        self.segment().normalize().classify()
+    }
+
+    fn reconstruct(&self) -> ReconstructedTokenIter<'_, '_, Vec<u8>> {
+        ReconstructedTokenIter { token_iter: self.tokenize(), original: self }
+    }
+}
+
+/// Structure used to tokenize a text with customs configurations.
```
Comment (Kerollmops) [inline src/tokenizer.rs:93]:
```suggestion
/// Structure used to tokenize a text with custom configurations.
```

File: src/tokenizer.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+/// let stop_words = Set::from_iter(["the"].iter()).unwrap();
+///
+/// // configurate stop words.
+/// let builder = builder.stop_words(&stop_words);
+///
+/// // build the tokenizer.
+/// let tokenizer = builder.build();
+/// ```
+///
+pub struct TokenizerBuilder<'o, 'sw, A: AsRef<[u8]>> {
```
Comment (Kerollmops) [inline src/tokenizer.rs:138]:
```suggestion
pub struct TokenizerBuilder<'o, 'sw, A> {
```

File: src/tokenizer.rs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    ///
+    /// # Arguments
+    ///
+    /// * `original` - the text to tokenize.
+    pub fn new(original: &'o str) -> TokenizerBuilder<'o, 'sw, Vec<u8>> {
+        Self { original, stop_words: None }
+    }
+}
+
+impl<'o, 'sw, A: AsRef<[u8]>> TokenizerBuilder<'o, 'sw, A> {
```
Comment (Kerollmops) [inline src/tokenizer.rs:154]:
```suggestion
impl<'o, 'sw, A> TokenizerBuilder<'o, 'sw, A> {
```

File: src/tokenizer.rs [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+    fn reconstruct(&self) -> ReconstructedTokenIter<'_, '_, Vec<u8>> {
+        ReconstructedTokenIter { token_iter: self.tokenize(), original: self }
+    }
+}
+
+/// Structure used to tokenize a text with customs configurations.
+///
+/// See [`TokenizerBuilder`] to know how to build a [`Tokenizer`].
+pub struct Tokenizer<'o, 'sw, A: AsRef<[u8]>> {
```
Comment (Kerollmops) [inline src/tokenizer.rs:96]:
```suggestion
pub struct Tokenizer<'o, 'sw, A> {
```

File: src/classifier/mod.rs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+/// Trait defining methods to classify [`Token`]s.
+pub trait Classify<'o>: Iterator
+where
+    Self: Sized,
+    Self: Iterator<Item = Token<'o>> + 'o,
+{
+    /// Assign to each [`Token`]s a [`TokenKind`].
+    ///
+    /// [`TokenKind`]: crate::TokenKind
```
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply src/classifier/mod.rs:30]: Yes, we can't remove it

File: src/segmenter/dummy.example.rs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+//     prefer JapaneseSegmenter (based on the Language) instead of LinderaSegmenter (based on the used Library).
+//     Same for the filename, prefer `japanese.rs` instead of `lindera.rs`.
+pub struct DummySegmenter;
+
+//TIP: All specialized segmenters only need to implement the method `segment_str` of the `Segmenter` trait.
+impl Segmenter for DummySegmenter {
+    fn segment_str<'a>(&self, to_segment: &'a str) -> Box<dyn Iterator<Item = &'a str> + 'a> {
+        //TIP: Create the iterator that will segment the provided text,
+        //     Here it returns the complete text without segmenting it.
+        let segment_iterator = Some(to_segment).into_iter();
```
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply src/segmenter/dummy.example.rs:20]: good idea.

File: src/detection.rs [hunk 8]
```diff
@@ -56,3 +56,193 @@ pub fn classify_separator(c: char) -> Option<SeparatorKind> {
...
+    Khm,
+    Tuk,
+    Aka,
+    Zul,
+    Sna,
+    Afr,
+    Lat,
+    Slk,
+    Cat
+    // Tgl  added in whatlang 0.13.0
```
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply src/detection.rs:151]: `Tgl` is Tagalog language we may update whatlang to have this enum variant.

File: src/detection.rs [hunk 9]
```diff
@@ -56,3 +56,193 @@ pub fn classify_separator(c: char) -> Option<SeparatorKind> {
...
+}
+
+pub struct StrDetection<'a> {
+    pub(crate) inner: &'a str,
+    pub(crate) script: Option<Script>,
+    pub(crate) language: Option<Language>,
+}
+
+impl<'a> StrDetection<'a> {
+    pub fn new(inner: &'a str) -> Self {
```
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply src/detection.rs]: This one is mandatory

**Review by Kerollmops — APPROVED** [review]
I like this PR and validate it!
bors merge

**Pull request merged**